README.md
<img src="./docs/imgs/kidronekit.png" width="100%">

# kidronekit

Premium extended version of dronekit

## Install

```
pip install kidronekit
```

## Getting Started

How to control get drone telemetrys via Mavlink with `kidronekit`:

```python
from kidronekit import connect, VehicleMode

connection_string = "tcp:192.168.0.120:5763"

# Connect to the Vehicle.
print("Connecting to vehicle on: %s" % (connection_string,))
vehicle = connect(connection_string, wait_ready=True)


# Get some vehicle attributes (state)
print("====================================================")
print(" GPS Status: %s" % vehicle.gps_0)
print(" Lat, Lon, Alt: %s" % vehicle.location.global_relative_frame.lat, vehicle.location.global_relative_frame.lon, vehicle.location.global_relative_frame.alt)
print(" Attitude: %s" % vehicle.attitude)
print(" Groundspeed: %s" % vehicle.groundspeed)    # settable
print(" Airspeed: %s" % vehicle.airspeed)    # settable
print(" Verticalspeed: %s" % vehicle.alt)
print(" Alt (MSL): %s" % vehicle.throttle) 
print(" Throttle: %s" % vehicle.climb)  
print(" Heading: %s" % vehicle.heading)
print(" Battery: %s" % vehicle.battery)
print(" Is Armable?: %s" % vehicle.is_armable)
print(" System status: %s" % vehicle.system_status.state)
print(" Mode: %s" % vehicle.mode.name)    # settable
print("====================================================")
    
```

Output:

```
====================================================
 GPS Status: GPSInfo:fix=6,num_sat=10
 Lat, Lon, Alt: 10.3939486 106.8843985 5.004
 Attitude: Attitude:pitch=-0.00046380647108890116,yaw=-1.9323097467422485,roll=-0.002339589409530163
 Groundspeed: 0.020496351644396782
 Airspeed: 0.0010000000474974513
 Verticalspeed: 17.920000076293945
 Alt (MSL): 0.020496351644396782
 Throttle: -0.005825158208608627
 Heading: 249
 Battery: Battery:voltage=12.587,current=28.16,level=0
 Is Armable?: True
 System status: ACTIVE
 Mode: GUIDED
====================================================
```

## License

Copyright (c) 3D Robotics and Kirinslab. All rights reserved. Licensed under the [Apache 2.0 License](LICENSE) license.

CHANGELOG.md
# Changelog

## Version 1.0.0 (2023-04-25)

### Initial
* Initial kidronekit from [Dronekit](https://github.com/dronekit/dronekit-python.git)

### Cleanup
* Removed some legacy and keep the core of the source code

### Improvements
* Support full data of `VFR_HUD` message from `Mavlink`: `airspeed`, `groundspeed`, `heading`, `throttle`, `alt`, `climb`


test.py
import time
from kidronekit import connect, VehicleMode

connection_string = "tcp:192.168.0.120:5763"

# Connect to the Vehicle.
print("Connecting to vehicle on: %s" % (connection_string,))
vehicle = connect(connection_string, wait_ready=True)

while True:
    # Get some vehicle attributes (state)
    print("====================================================")
    print(" GPS Status: %s" % vehicle.gps_0)
    print(" Lat, Lon, Alt: %s" % vehicle.location.global_relative_frame.lat, vehicle.location.global_relative_frame.lon, vehicle.location.global_relative_frame.alt)
    print(" Attitude: %s" % vehicle.attitude)
    print(" Groundspeed: %s" % vehicle.groundspeed)    # settable
    print(" Airspeed: %s" % vehicle.airspeed)    # settable
    print(" Verticalspeed: %s" % vehicle.alt)
    print(" Alt (MSL): %s" % vehicle.throttle) 
    print(" Throttle: %s" % vehicle.climb)  
    print(" Heading: %s" % vehicle.heading)
    print(" Battery: %s" % vehicle.battery)
    print(" Is Armable?: %s" % vehicle.is_armable)
    print(" System status: %s" % vehicle.system_status.state)
    print(" Mode: %s" % vehicle.mode.name)    # settable
    print("====================================================")
    time.sleep(0.5)

kidronekit/mavlink.py
from __future__ import print_function

import logging
import time
import socket
import errno
import sys
import os
import platform
import copy
from kidronekit import APIException
from pymavlink import mavutil
from queue import Queue, Empty
from threading import Thread

if platform.system() == 'Windows':
    from errno import WSAECONNRESET as ECONNABORTED
else:
    from errno import ECONNABORTED


class MAVWriter(object):
    """
    Indirection layer to take messages written to MAVlink and send them all
    on the same thread.
    """

    def __init__(self, queue):
        self._logger = logging.getLogger(__name__)
        self.queue = queue

    def write(self, pkt):
        self.queue.put(pkt)

    def read(self):
        self._logger.critical('writer should not have had a read request')
        os._exit(43)


class mavudpin_multi(mavutil.mavfile):
    '''a UDP mavlink socket'''
    def __init__(self, device, baud=None, input=True, broadcast=False, source_system=255, source_component=0, use_native=mavutil.default_native):
        self._logger = logging.getLogger(__name__)
        a = device.split(':')
        if len(a) != 2:
            self._logger.critical("UDP ports must be specified as host:port")
            sys.exit(1)
        self.port = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.udp_server = input
        self.broadcast = False
        self.addresses = set()
        if input:
            self.port.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            self.port.bind((a[0], int(a[1])))
        else:
            self.destination_addr = (a[0], int(a[1]))
            if broadcast:
                self.port.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
                self.broadcast = True
        mavutil.set_close_on_exec(self.port.fileno())
        self.port.setblocking(False)
        mavutil.mavfile.__init__(self, self.port.fileno(), device, source_system=source_system, source_component=source_component, input=input, use_native=use_native)

    def close(self):
        self.port.close()

    def recv(self, n=None):
        try:
            try:
                data, new_addr = self.port.recvfrom(65535)
            except socket.error as e:
                if e.errno in [errno.EAGAIN, errno.EWOULDBLOCK, errno.ECONNREFUSED]:
                    return ""
            if self.udp_server:
                self.addresses.add(new_addr)
            elif self.broadcast:
                self.addresses = {new_addr}
            return data
        except Exception:
            self._logger.exception("Exception while reading data", exc_info=True)

    def write(self, buf):
        try:
            try:
                if self.udp_server:
                    for addr in self.addresses:
                        self.port.sendto(buf, addr)
                else:
                    if len(self.addresses) and self.broadcast:
                        self.destination_addr = self.addresses[0]
                        self.broadcast = False
                        self.port.connect(self.destination_addr)
                    self.port.sendto(buf, self.destination_addr)
            except socket.error:
                pass
        except Exception:
            self._logger.exception("Exception while writing data", exc_info=True)

    def recv_msg(self):
        '''message receive routine for UDP link'''
        self.pre_message()
        s = self.recv()
        if len(s) > 0:
            if self.first_byte:
                self.auto_mavlink_version(s)

        m = self.mav.parse_char(s)
        if m is not None:
            self.post_message(m)

        return m


class MAVConnection(object):

    def stop_threads(self):
        if self.mavlink_thread_in is not None:
            self.mavlink_thread_in.join()
            self.mavlink_thread_in = None
        if self.mavlink_thread_out is not None:
            self.mavlink_thread_out.join()
            self.mavlink_thread_out = None

    def __init__(self, ip, baud=115200, target_system=0, source_system=255, source_component=0, use_native=False):
        self._logger = logging.getLogger(__name__)

        if ip.startswith("udpin:"):
            self.master = mavudpin_multi(ip[6:], input=True, baud=baud, source_system=source_system, source_component=source_component)
        else:
            self.master = mavutil.mavlink_connection(ip, baud=baud, source_system=source_system, source_component=source_component)

        # TODO get rid of "master" object as exposed,
        # keep it private, expose something smaller for kidronekit
        self.out_queue = Queue()
        self.master.mav = mavutil.mavlink.MAVLink(
            MAVWriter(self.out_queue),
            srcSystem=self.master.source_system,
            srcComponent=self.master.source_component,
            use_native=use_native)

        # Monkey-patch MAVLink object for fix_targets.
        sendfn = self.master.mav.send

        def newsendfn(mavmsg, *args, **kwargs):
            self.fix_targets(mavmsg)
            return sendfn(mavmsg, *args, **kwargs)

        self.master.mav.send = newsendfn

        # Targets
        self.target_system = target_system

        # Listeners.
        self.loop_listeners = []
        self.message_listeners = []

        # Debug flag.
        self._accept_input = True
        self._alive = True
        self._death_error = None

        import atexit

        def onexit():
            self._alive = False
            self.stop_threads()

        atexit.register(onexit)

        def mavlink_thread_out():
            # Huge try catch in case we see http://bugs.python.org/issue1856
            try:
                while self._alive:
                    try:
                        msg = self.out_queue.get(True, timeout=0.01)
                        self.master.write(msg)
                    except Empty:
                        continue
                    except socket.error as error:
                        # If connection reset (closed), stop polling.
                        if error.errno == ECONNABORTED:
                            raise APIException('Connection aborting during read')
                        raise
                    except Exception as e:
                        self._logger.exception('mav send error: %s' % str(e))
                        break
            except APIException as e:
                self._logger.exception("Exception in MAVLink write loop", exc_info=True)
                self._alive = False
                self.master.close()
                self._death_error = e

            except Exception as e:
                # http://bugs.python.org/issue1856
                if not self._alive:
                    pass
                else:
                    self._alive = False
                    self.master.close()
                    self._death_error = e

            # Explicitly clear out buffer so .close closes.
            self.out_queue = Queue()

        def mavlink_thread_in():
            # Huge try catch in case we see http://bugs.python.org/issue1856
            try:
                while self._alive:
                    # Loop listeners.
                    for fn in self.loop_listeners:
                        fn(self)

                    # Sleep
                    self.master.select(0.05)

                    while self._accept_input:
                        try:
                            msg = self.master.recv_msg()
                        except socket.error as error:
                            # If connection reset (closed), stop polling.
                            if error.errno == ECONNABORTED:
                                raise APIException('Connection aborting during send')
                            raise
                        except mavutil.mavlink.MAVError as e:
                            # Avoid
                            #   invalid MAVLink prefix '73'
                            #   invalid MAVLink prefix '13'
                            self._logger.debug('mav recv error: %s' % str(e))
                            msg = None
                        except Exception:
                            # Log any other unexpected exception
                            self._logger.exception('Exception while receiving message: ', exc_info=True)
                            msg = None
                        if not msg:
                            break

                        # Message listeners.
                        for fn in self.message_listeners:
                            try:
                                fn(self, msg)
                            except Exception:
                                self._logger.exception(
                                    'Exception in message handler for %s' % msg.get_type(),
                                    exc_info=True
                                )

            except APIException as e:
                self._logger.exception('Exception in MAVLink input loop')
                self._alive = False
                self.master.close()
                self._death_error = e
                return

            except Exception as e:
                # http://bugs.python.org/issue1856
                if not self._alive:
                    pass
                else:
                    self._alive = False
                    self.master.close()
                    self._death_error = e

        t = Thread(target=mavlink_thread_in)
        t.daemon = True
        self.mavlink_thread_in = t

        t = Thread(target=mavlink_thread_out)
        t.daemon = True
        self.mavlink_thread_out = t

    def reset(self):
        self.out_queue = Queue()
        if hasattr(self.master, 'reset'):
            self.master.reset()
        else:
            try:
                self.master.close()
            except:
                pass
            self.master = mavutil.mavlink_connection(self.master.address)

    def fix_targets(self, message):
        """Set correct target IDs for our vehicle"""
        if hasattr(message, 'target_system'):
            message.target_system = self.target_system

    def forward_loop(self, fn):
        """
        Decorator for event loop.
        """
        self.loop_listeners.append(fn)

    def forward_message(self, fn):
        """
        Decorator for message inputs.
        """
        self.message_listeners.append(fn)

    def start(self):
        if not self.mavlink_thread_in.is_alive():
            self.mavlink_thread_in.start()
        if not self.mavlink_thread_out.is_alive():
            self.mavlink_thread_out.start()

    def close(self):
        # TODO this can block forever if parameters continue to be added
        self._alive = False
        while not self.out_queue.empty():
            time.sleep(0.1)
        self.stop_threads()
        self.master.close()

    def pipe(self, target):
        target.target_system = self.target_system

        # vehicle -> self -> target
        @self.forward_message
        def callback(_, msg):
            try:
                target.out_queue.put(msg.pack(target.master.mav))
            except:
                try:
                    assert len(msg.get_msgbuf()) > 0
                    target.out_queue.put(msg.get_msgbuf())
                except:
                    self._logger.exception('Could not pack this object on receive: %s' % type(msg), exc_info=True)

        # target -> self -> vehicle
        @target.forward_message
        def callback(_, msg):
            msg = copy.copy(msg)
            target.fix_targets(msg)
            try:
                self.out_queue.put(msg.pack(self.master.mav))
            except:
                try:
                    assert len(msg.get_msgbuf()) > 0
                    self.out_queue.put(msg.get_msgbuf())
                except:
                    self._logger.exception('Could not pack this object on forward: %s' % type(msg), exc_info=True)

        return target


examples/telemetrys/telemetrys.py
import time
from kidronekit import connect, VehicleMode

connection_string = "tcp:192.168.0.120:5763"

# Connect to the Vehicle.
print("Connecting to vehicle on: %s" % (connection_string,))
vehicle = connect(connection_string, wait_ready=True)

while True:
    # Get some vehicle attributes (state)
    print("====================================================")
    print(" GPS Status: %s" % vehicle.gps_0)
    print(" Lat, Lon, Alt: %s" % vehicle.location.global_relative_frame.lat, vehicle.location.global_relative_frame.lon, vehicle.location.global_relative_frame.alt)
    print(" Attitude: %s" % vehicle.attitude)
    print(" Groundspeed: %s" % vehicle.groundspeed)    # settable
    print(" Airspeed: %s" % vehicle.airspeed)    # settable
    print(" Verticalspeed: %s" % vehicle.climb) 
    print(" Heading: %s" % vehicle.heading)
    print(" Battery: %s" % vehicle.battery)
    print(" Is Armable?: %s" % vehicle.is_armable)
    print(" System status: %s" % vehicle.system_status.state)
    print(" Mode: %s" % vehicle.mode.name)    # settable
    print("====================================================")
    time.sleep(0.5)